url49584c1d41_Arrem_GGS_tgz-pJ8-com_gamezgalaxy_test_console_MainJ8/src/com/gamezgalaxy/GGS/world/Level.java

```diff
@@ -0,0 +1,424 @@
+        if (x >= width) { return -1; }
+        if (y < 0) { return -1; }
+        if (y >= height) { return -1; }
+        if (z < 0) { return -1; }
+        if (z >= depth) { return -1; }
+        return x + z * width + y * width * depth;
+    }
+	
+	public int[] IntToPos(int index) {
+		int[] toreturn = new int[3];
+		toreturn[1] = (index / width / height);
+		index -= toreturn[1]*width*height;
+		toreturn[2] = (index/width);
+		index -= toreturn[2]*width;
+		toreturn[0] = index;
+		return toreturn;
+	}
+	
+	/**
+	 * Save the level
+	 * @throws IOException
+	 *                   An IOExceptoin is thrown if there is a problem writing to the file
+	 */
+	public void Save() throws IOException {
+		if (!new File("levels").exists())
+			new File("levels").mkdir();
+		FileOutputStream fos = new FileOutputStream("levels/" + name + ".ggs");
+		GZIPOutputStream gos = new GZIPOutputStream(fos);
+		ObjectOutputStream out = new ObjectOutputStream(gos);
+		out.writeLong(serialVersionUID);
+		out.writeObject(this);
+		out.close();
+		gos.close();
+		fos.close();
+	}
+	
+	/**
+	 * Unload this level.
+	 * All players who are in this level will be sent to the {@link Server#MainLevel}
+	 * @param server
+	 *             The server thats unloading the level
+	 * @throws IOException
+	 *                   An IOException will occur if there is a problem saving the level
+	 */
+	public void Unload(Server server) throws IOException {
+		Unload(server, true);
+	}
+	/**
+	 * Unload this level.
+	 * All players who are in this level will be sent to the {@link Server#MainLevel}
+	 * @param server
+	 *             The server thats unloading the level
+	 * @param save
+	 *           Weather the level should save before unloading
+	 * @throws IOException
+	 *                   An IOException will occur if there is a problem saving the level
+	 */
+	public void Unload(Server server, boolean save) throws IOException {
+		if (save)
+			Save();
+		run = false;
+		try {
+			physics.join();
+		} catch (InterruptedException e) {
+			e.printStackTrace();
+		}
+		server.Log("Unloading " + name);
+		for (Player p : server.players) {
+			if (p.getLevel() == this)
+				p.changeLevel(server.MainLevel, false);
+		}
+		blocks = null;
+	}
+	
+	/**
+	 * Load a level and return the level
+	 * @param filename
+	 *               The file to load and read
+	 * @return
+	 *        The level object
+	 * @throws IOException
+	 *                   An IOException is thrown if there is a problem reading the level
+	 * @throws ClassNotFoundException
+	 *                              This exception is thrown if a block that was saved with the level is not loaded or cant be found.
+	 */
+	public static Level Load(String filename) throws IOException, ClassNotFoundException {
+		Level l = null;
+		if (filename.endsWith(".dat"))
+			l = Convert(filename);
+		else {
+			FileInputStream fis = new FileInputStream(filename);
+			GZIPInputStream gis = new GZIPInputStream(fis);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *gis*. The resource is referenced at lines 331, 332, 333, 335. The resource is closed at lines 337, 345, 346, 347. There are other execution paths that do not contain closure statements, for example, when ObjectInputStream.readLong throws an exception. Ensure *gis* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).